More IR mini passes

Akirathan opened the pull request:
Closes #11326

### Pull Request Description

Mega passes  converted to minipasses in `Passes.globalTypingPasses` group:
https://github.com/enso-org/enso/blob/4e6837f9dad4b8baaf1140774126348c662c5388/engine/runtime-compiler/src/main/scala/org/enso/compiler/Passes.scala#L41-L63

Are:
- [MethodDefinitions](https://github.com/enso-org/enso/blob/a3321fee972bcb2b9f3292286a97708988fc07d9/engine/runtime-compiler/src/main/java/org/enso/compiler/pass/resolve/MethodDefinitions.java)
- [ImportSymbolAnalysis](https://github.com/enso-org/enso/blob/8d9db8fe01b70b2992103c446cac31bc266c10ff/engine/runtime-compiler/src/main/java/org/enso/compiler/pass/analyse/ImportSymbolAnalysis.java)
- [AmbiguousImportsAnalysis](https://github.com/enso-org/enso/blob/8d9db8fe01b70b2992103c446cac31bc266c10ff/engine/runtime-compiler/src/main/java/org/enso/compiler/pass/analyse/AmbiguousImportsAnalysis.java)
- [PrivateModuleAnalysis](https://github.com/enso-org/enso/blob/8d9db8fe01b70b2992103c446cac31bc266c10ff/engine/runtime-compiler/src/main/java/org/enso/compiler/pass/analyse/PrivateModuleAnalysis.java)
- [PrivateConstructorAnalysis](https://github.com/enso-org/enso/blob/303c9fe545c7968d57810e39499c40da357075f0/engine/runtime-compiler/src/main/java/org/enso/compiler/pass/analyse/PrivateConstructorAnalysis.java)

All of these mini passes are ordered after each other. So now, in `Passes.globalTypingPasses` group, there are **8** consecutive minipasses.

### Important Notes

<!--
- Mention important elements of the design.
- Mention any notable changes to APIs.
-->

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.
- [x] Benchmarks are solid:
    - [x] engine benchmarks [run #1](https://github.com/enso-org/enso/actions/runs/12044380533)
      - In run 1, MethodDefinitions is still mega pass. Scheduling another one:
    - [x] engine benchmarks [run #2](https://github.com/enso-org/enso/actions/runs/12055070861)

Commit:
Move ScalaConversions from runtime to engine-common

Commit:
Convert ImportSymbolAnalysis to mini pass

Commit:
fmt

JaroslavTulach commented on engine/runtime-compiler/src/main/java/org/enso/compiler/pass/analyse/ImportSymbolAnalysis.java:
```diff
@@ -0,0 +1,194 @@
+
+    @Override
+    public Expression transformExpression(Expression expr) {
+      return expr;
```
This is an empty, useless call that might slow the traversal down (on the instruction level). Didn't we wanted to have a way to let the `MiniIRPass` to instruct the infrastructure to skip processing of subtrees? Shouldn't this mini pass:
```java
  public MiniIRPass prepare(IR parent, Expression child) {
    return null;
  }
```
e.g. remove itself from any `Expression` processing? In such case the `transformExpression` should `throw new IllegalStateException()` as it should be never called.

Akirathan replied:
The problem is that this pass only wishes to transform `Module` IR. More specifically, it only does something useful in `transformModule`. And `transformModule` is called as the last one during the traversal - because it transforms the root node. If you would tell the traversing system that you wish to skip the rest of the iteration in any expression, you would never reach the Module root node.

JaroslavTulach replied:
> would never reach the Module root node.

That's not how the system is supposed to work and I would be surprised if it worked that way:
- `MiniPassTraverser.compileDeep` receives `IR.Module`
- it creates `rootTask` for the `IR.Module` object
- it puts the task to `stackOfPendingIrs`
- then something happens
- at the end `rootTask.convertExpression` is called
- inside of `converExpresison` as call to `transformModule` is made

The [chaining code](https://github.com/enso-org/enso/blob/a6460c17b27eb40d90a64577e8192400d76ec678/engine/runtime-compiler/src/main/java/org/enso/compiler/pass/ChainedMiniPass.java#L28) is probably not ready for `null` from `prepare` being returned, but we could (and should) adjust it to that for the sake of eliminating needless traversal.

Akirathan replied:
Traversal for this mini pass is skipped in 806406cbd567047ef3db73a84c7c413daf7de9cc

JaroslavTulach commented on build.sbt:
```diff
@@ -2118,6 +2118,7 @@ lazy val `engine-common` = project
   .enablePlugins(JPMSPlugin)
   .settings(
     frgaalJavaCompilerSetting,
+    scalaModuleDependencySetting,
```
The comment for this module says:

> engine/common project contains classes that are necessary to configure GraalVM's polyglot context. Most specifically it contains `ContextFactory`. As such it needs to depend on `org.graalvm.polyglot` package. Otherwise its dependencies shall be limited - no JSON & co. please. 

This change adds a dependency on Scala. Knowing we want to get rid of Scala this probably contra-productive (effectively all our `runtime*` projects will have transitive Scala dependency regardless whether they need it or not). Isn't there some other module that already depends on Scala where to put `ScalaConversion` utility?

Akirathan replied:
Cannot think of one. `ScalaConversions` is used by code from `runtime`, `runtime-compiler`, `runtime-instrument-common` and `runtime-integrations-test`. The only project reachable from all of those is `engine-common`. The tests are green now. Let's not mess with this in this PR and address it once we will be dealing with dependencies again.

JaroslavTulach replied:
Famous words: let's fix it later. I don't think I agree with that. 

> ScalaConversions is used by code from runtime, runtime-compiler, runtime-instrument-common and runtime-integrations-test. The only project reachable from all of those is engine-common

Finding a common dependency module, moving some utility method there, adding a dependency to it _ e.g. making it _heavier_ - is a known _modular fallacy_. Just don't do that.

Moreover there is a better module common to the listed ones `scala-libs-wrapper`. Put the common Scala utilities there:
- they will be shared
- they will be co-located with Scala
- they will disappear if Scala wrapper is no longer needed

Akirathan replied:
EDIT: It seems that it is possible to move `ScalaConversions` to `scala-libs-wrapper`. I will try that.

Akirathan replied:
Moved to `scala-libs-wrapper` in 417843dc850d49075f89970f4284749e15118e5d

JaroslavTulach reviewed (commented):
I would rather not increase the amount of projects that depend on Scala.

Commit:
createForInlineCompilation cannot return null

Commit:
ChainedMiniPass delegates to transformModule

Commit:
ImportSymbolAnalysis copies unchanged imports

Commit:
Fix compilation of a test

Commit:
Combining mini passes can result in null

Commit:
Merge branch 'develop' into wip/akirathan/11326-more-mini-passes

Commit:
Add consistency validation to PassManager

Commit:
Add assert that bindingsMap != null

Commit:
Consistency validation has better error message.

Commit:
CompilerTestSetup explicitly changes IR in ModuleContext.

Commit:
runtime-integration-tests accesses PassManager field via reflection

Commit:
Add MiniPassTraverserTest

Commit:
MiniPassTraverser supports end of the traversal

Commit:
Fix MockMiniPass - stopExpression is not even prepared

Commit:
ImportSymbolAnalysis skips processing

Commit:
Merge branch 'develop' into wip/akirathan/11326-more-mini-passes

JaroslavTulach commented on engine/runtime-compiler/src/main/java/org/enso/compiler/pass/analyse/ImportSymbolAnalysis.java:
```diff
@@ -86,9 +87,16 @@ public Module transformModule(Module moduleIr) {
     @Override
     public Expression transformExpression(Expression expr) {
-      return expr;
+      throw new IllegalStateException("Should not be called - prepare returns null");
```
Might it be called for _inline processing_?

Akirathan replied:
Nope - 2bcdede83dc8a02e2bb9e32eb0128f5cbb81ec33. ImportSymbolAnalysis does not make sense for inline compilation - it only cares about the top level `org.enso.compiler.core.ir.Module` element.

Commit:
Migrate AmbiguousImportsAnalysis to mini pass and to Java

Commit:
Add some tests for chained passes

Commit:
ChainedMiniPass handles null passes

Commit:
Fix compilation error in PasesTest

Commit:
Merge branch 'develop' into wip/akirathan/11326-more-mini-passes

Commit:
Fix NPE

Commit:
Fix errors in polyglot imports

Commit:
Remove unnecessary isntancdeof

Commit:
MiniPassTest does not call PassManager.runModule directly.

This violates the consistency between IR from ModuleContext.

Commit:
Move tests from runtime-compiler to runtime-integration-tests

Commit:
MockMiniPass uses builder pattern

Commit:
Remove EventRecorder, MockMiniPass is not IRProcessingPass

Commit:
Move MiniPassTraverserTest back into runtime-compiler/Test

Commit:
PassManager.runInline also chains mini passes.

Make the algorithm for running passes generic

Commit:
MiniPassFactory.createForInlineCompilation can return null.

JaroslavTulach commented on engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/PassManager.scala:
```diff
@@ -198,27 +153,97 @@ class PassManager(
+    *                     Either [[ModuleContext]] or [[InlineContext]]
+    * @return Compiled IR. Might be the same reference as `ir` if no compilation was done.
+    */
+  private def runPasses[IRType <: IR, ContextType](
```
One method to do the loop used from both `runPassesOnModule` and `runPassesInline` is a great improvement! I wanted it, but I got scared by the necessary typings when [combining subsequent minipasses](https://github.com/enso-org/enso/pull/11191/commits/2736a7642886c9a8919fb037ef21463422a2c001).

Commit:
Fix IR inconsistency in TypeInferenceTest

Commit:
Remove unused method from PassManager

Commit:
ImportSymbolAnalysis does not support inline compilation.

This fixes std-benchmarks/compile.

Commit:
Convert PrivateModuleAnalysis to mini pass

Commit:
Fix spurious errors

Commit:
Convert PrivateConstructorAnalysis to mini pass

Commit:
FIx compilation error in a test

Commit:
Convert MethodDefinitions to Java

Commit:
FIx compilation error in a test

Commit:
Fix NPE

Commit:
FIx compilation error in a test

Commit:
MetadataStorage.EMPTY is not a global reference

Commit:
FIx compilation error in a test

Commit:
MethodDefinitions.lambdaWithNewSelfArg copies arguments

Commit:
Fix copy of static method in MethodDefinitions

Commit:
Ensure consistency between ModuleContext and IR in uncachedParseModule

Commit:
Better error message in IrToTruffle.generateReExportBindings

Commit:
Fix No_Such_Method error.

Problem was in MetadataStorage.get(MethodDefinitions.INSTANCE) that returned null if MethodDefinitions is not final.

Commit:
Merge branch 'develop' into wip/akirathan/11326-more-mini-passes

Commit:
Fix invalid IR in GlobalNamesTest

JaroslavTulach commented on engine/runtime-compiler/src/main/java/org/enso/compiler/pass/resolve/MethodDefinitions.java:
```diff
@@ -0,0 +1,338 @@
+                });
+
+    java.util.List<Definition> withStaticAliases = new ArrayList<>();
+    for (var def : CollectionConverters.asJava(newDefs)) {
```
I guess the above `.map` body and the `for` loop body could easily be merged together into one big `for` loop.

JaroslavTulach commented on engine/runtime-compiler/src/main/java/org/enso/compiler/pass/resolve/MethodDefinitions.java:
```diff
@@ -0,0 +1,338 @@
+   * @param method Non-static method from which a static alias method is generated.
+   * @return Static alias method for the given {@code method} or null.
+   */
+  private Method.Explicit generateStaticAliasMethod(Method.Explicit method) {
```
Unrelated to this PR.

What would it take to **register only static methods** in each type? What's an invocation of `(x:Integer).to_text` anyway? It should be convertible to `Integer.to_text x` - e.g. an instance invocation just resolves the right method based on the type of the instances, but then it invokes the static `Function`.

Benefit:
- reduce the amount of `Function` instances in the system by half

Prior art:
- Graal compiler sees all JVM methods as static (instances methods of a class being static and having `this` as first argument)
- the compiler only has special `InvokeNode` that deals with finding the right method to invoke when virtual dispatch is necessary

Akirathan replied:
Created an issue https://github.com/enso-org/enso/issues/11686 from your comment.

Commit:
Convert MethodDefinitions to mini pass

Commit:
FIx compilation of MethodDefinitionsTest

Commit:
Remove TODO comment

JaroslavTulach commented on engine/runtime-compiler/src/main/java/org/enso/compiler/pass/ChainedMiniPass.java:
```diff
@@ -15,37 +15,43 @@ private ChainedMiniPass(MiniIRPass firstPass, MiniIRPass secondPass) {
   static MiniIRPass chain(MiniIRPass firstPass, MiniIRPass secondPass) {
-    if (firstPass == null) {
-      return secondPass;
-    }
```
Why the removal? I believe we want the `chain` method to return `null` when neither of the passes wants to process anything? E.g.
- one of the passes is `null` - return the _other_
- both are `null` return `null`
- otherwise wrap them by `ChainedMiniPass`

Akirathan replied:
Reintroduced again in 273bea8f37c724ffa9ae424af0b2314707af4ac4

JaroslavTulach commented on engine/runtime-compiler/src/main/java/org/enso/compiler/pass/ChainedMiniPass.java:
```diff
@@ -15,37 +15,43 @@ private ChainedMiniPass(MiniIRPass firstPass, MiniIRPass secondPass) {
       return this;
     } else {
-      return new ChainedMiniPass(first, second);
+      return new ChainedMiniPass(firstPrepared, secondPrepared);
```
This should call `chain` (or `combine`) method to allow returning `null`. Otherwise a group of chained passes cannot disable itself completely from processing of the subtree.

Akirathan replied:
Good catch. Fixed in bd10bf6d84df4e1442495b8d5cd81bfe0b0c6069

JaroslavTulach commented on engine/runtime-compiler/src/main/java/org/enso/compiler/pass/ChainedMiniPass.java:
```diff
@@ -15,37 +15,43 @@ private ChainedMiniPass(MiniIRPass firstPass, MiniIRPass secondPass) {
 
+  @Override
+  public Module transformModule(Module moduleIr) {
+    var first = firstPass == null ? moduleIr : firstPass.transformModule(moduleIr);
```
For consistency with `transformExpression` method call the variable `fstIr`, maybe?

Akirathan replied:
Reasonable renaming is always a good idea. Done in b3f750eb52704a386b26400f54cc8d685ff3bc60

JaroslavTulach commented on engine/runtime-compiler/src/main/java/org/enso/compiler/pass/MiniIRPass.java:
```diff
@@ -103,10 +105,17 @@ public String toString() {
+    }
+    if (second == null) {
+      return first;
+    }
```
Oh, the check for `null` is here. I was expecting it to be inside of `static MiniIRPass chain` - that was the whole point of having a `factory` method - otherwise one can use just the constructor directly, but one has to use `combine` method consistently wherever allocating new chain then.

Akirathan replied:
Check for null moved to the factory method in 273bea8f37c724ffa9ae424af0b2314707af4ac4

JaroslavTulach commented on engine/runtime-compiler/src/main/java/org/enso/compiler/pass/analyse/ImportSymbolAnalysis.java:
```diff
@@ -0,0 +1,210 @@
+ * the IR import with {@link org.enso.compiler.core.ir.expression.errors.ImportExport}. Reports only
+ * the first unresolved symbol.
+ */
+public class ImportSymbolAnalysis implements MiniPassFactory {
```
```suggestion
public final class ImportSymbolAnalysis implements MiniPassFactory {
```

Akirathan replied:
Dione in cb5f8c1b6c194b1c8c5b255646d6d044e7389745

JaroslavTulach commented on engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/PassManager.scala:
```diff
@@ -18,7 +22,7 @@ import org.enso.compiler.core.CompilerError
 //noinspection DuplicatedCode
 class PassManager(
-  passes: List[PassGroup],
+  protected val passes: List[PassGroup],
```
`protected`!? Is there a subclass of `PassManager`?

Akirathan replied:
Made private in 3e1ade9fff64052528ddc436213675b14c458bc2 . Note that it has to stay as field - we need it to be a field so that it is accessible in tests via https://github.com/enso-org/enso/blob/fb3e29d87b1328ca8bb78ef73c76421cc1e6db8f/engine/runtime-integration-tests/src/test/scala/org/enso/compiler/test/CompilerTest.scala#L34-L46

Akirathan replied:
EDIT: It has to be protected as it is accessed in tests in [PassManagerTestUtils](https://github.com/enso-org/enso/blob/546f17cde22689d78255def45038d13a96d25ebe/engine/runtime-instrument-common/src/test/java/org/enso/compiler/pass/PassManagerTestUtils.java#L9).

JaroslavTulach commented on engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/PassManager.scala:
```diff
@@ -231,6 +247,51 @@ class PassManager(
   }
 
+  /** Validates consistency between the IR accessible via `moduleContext` and `ir`.
+    * There is no way how to enforce this consistency statically.
```
```suggestion
    * There is no way to enforce this consistency statically.
```

Akirathan replied:
Typo fixed in 85ab82f3ce9891445b4cd4ccddbbdff7361abf51

JaroslavTulach commented on engine/runtime-compiler/src/test/java/org/enso/compiler/test/pass/MiniPassTraverserTest.java:
```diff
@@ -0,0 +1,108 @@
+    var miniPass = MockMiniPass.builder().stopExpr(e3).build();
+    MiniIRPass.compile(MockExpression.class, e1, miniPass);
+    assertThat("e3 should not be processed", e3.isPreparedByAny(), is(false));
+    assertThat("e3 should not be processed", e3.isTransformedByAny(), is(false));
```
I am missing a test that would check a `ChainedMiniPass` is able to skip some elements in the tree - e.g. that there can be `!element.isTransformedByAny()` in an `IR` traversed by a chained mini pass. Reading the current code I assume such a test should fail.

Great that the behavior of `MiniPassTraverser` is tested at all!

Akirathan replied:
Added such a test in 023d1aa4165a805b7f5273209601f98435b9d4cc

Commit:
Move ScalaConversions to scala-libs-wrapper

JaroslavTulach reviewed (approved):
Thanks for moving the shared Scala utilities to `scala-libs-wrapper` and keeping `engine-common` as lightweight as it was. That was the biggest architectural flaw.

From a code point of view, I believe there is a mistake in `ChainedMiniPass` - it can never unregister completely from the traversal.

Otherwise it all looks good and thanks for increasing the test coverage.

Commit:
Move check for null minipasses to the factory method when combining

Commit:
ChainedMiniPass can disable itself from processing subtree

Commit:
Rename variables

Commit:
ImportSymbolAnalysis is final

Commit:
Typo in docs

Commit:
Make PassManager.passes field private

Commit:
Fix: ChainedMiniPass can disable itself from processing subtree

Commit:
Add mini pass traverser test

Commit:
PassManager.passes field needs to be protected

(package private)
